pages/api/utils/occurrence-utils.js

```diff
@@ -51,39 +53,69 @@ const matchAndMapVulnerabilities = (occurrences) => {
       occurrence.discovered.discovered.analysisStatus === "FINISHED_SUCCESS"
   );
 
-  return scanStarts.map((scan) => {
-    const startTime = dayjs(scan.createTime);
-    // get all scans that end after the start time
-    const possibleScanEnds = completedScans.filter((occurrence) =>
-      dayjs(occurrence.createTime).isSameOrAfter(startTime)
-    );
-    // pick the scan that ended closest to the start time
-    const scanEndTimes = possibleScanEnds.map((scan) =>
-      dayjs(scan.createTime).valueOf()
-    );
-    const closestEndTime = Math.min(...scanEndTimes);
-    const matchingScanEndOccurrence = possibleScanEnds.find(
-      (scan) => dayjs(scan.createTime).valueOf() === closestEndTime
-    );
+  const matchedOccurrences = scanStarts
+    .map((startScan) => {
+      const startTime = dayjs(startScan.createTime);
+      // get all scans that end after the start time
+      const possibleScanEnds = completedScans.filter((occurrence) =>
+        dayjs(occurrence.createTime).isSameOrAfter(startTime)
+      );
+      // pick the scan that ended closest to the start time
+      const scanEndTimes = possibleScanEnds.map((scan) =>
+        dayjs(scan.createTime).valueOf()
+      );
+      const closestEndTime = Math.min(...scanEndTimes);
+      const matchingScanEndOccurrence = possibleScanEnds.find(
+        (scan) => dayjs(scan.createTime).valueOf() === closestEndTime
+      );
+
+      if (!matchingScanEndOccurrence) {
+        unmatchedOccurrences.push(startScan);
+        return null;
+      }
+
+      const matchingVulnerabilityOccurrences = vulnerabilityOccurrences.filter(
+        (vuln) =>
+          vuln.kind === "VULNERABILITY" &&
+          vuln.createTime === matchingScanEndOccurrence?.createTime
+      );
+
+      return {
+        name: startScan.name,
+        started: startScan.createTime,
+        completed: matchingScanEndOccurrence?.createTime,
+        vulnerabilities: mapVulnerabilities(matchingVulnerabilityOccurrences),
+        originals: [
+          startScan,
+          matchingScanEndOccurrence,
+          ...matchingVulnerabilityOccurrences,
+        ],
+      };
+    })
```

Comment:
yeah I can see how this will get gross pretty fast if we don't have stricter requirements for occurrences on the API side, and aren't using notes to help tie occurrences together